docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,180 @@
 #!/usr/bin/env python3
 import argparse
+import hashlib
+import json
+import subprocess
+import typing
+
+
+class ImageNotFoundError(ValueError):
+    pass
+
+
+class _ImageKey(typing.NamedTuple):
+    commands_hash: str
+    packages_hash: str
+
+
+def _docker_push_latest_if_changed(source, target, *, is_dry_run):
+    _validate_source(source)
+    target = _get_sanitized_target(source, target)
+
+    print('Pushing source image')
+    _push_image(source, is_dry_run=is_dry_run)
+    try:
+        print('Pulling target image...')
+        _pull_image(target)
+    except ImageNotFoundError:
+        print((
+            f'Target image {target} was not found in the registry. '
+            'Going to attempt to tag and push the target image anyway.'
+        ))
+        _tag_image(source, target, is_dry_run=is_dry_run)
+        _push_image(target, is_dry_run=is_dry_run)
+    else:
+        if _has_image_changed(source, target):
+            print((
+                'Source image was found to be different from the current '
+                'version of the target image. Tagging and pushing new target '
+                'image.'
+            ))
+            _tag_image(source, target, is_dry_run=is_dry_run)
+            _push_image(target, is_dry_run=is_dry_run)
+        else:
+            print((
+                'Source image was found to be the same as the current '
+                'version of the target image. Not pushing the target image.'
+            ))
+
+
+def _validate_source(source):
+    print('Docker inspecting source to verify it exists')
+    source_inspect = _inspect_image(source)
+    if source not in source_inspect['RepoTags']:
+        # Hack to check if source is just a repo without a tag.
+        # An inspect query with just a repo will search for
+        # latest, but the result in RepoTags will only have
+        # repo:latest and not just repo
+        raise ValueError((
+            f'The source image {source} does not have a tag! '
+            'You must include a tag in the source parameter.'
+        ))
+
+
+def _get_sanitized_target(source, target):
+    if not target:
+        repository = source.rpartition(':')[0]
```

Comment:
hmmm, this might fail again for the same `port` reason you listed above.  Maybe this needs `urllib.parse` as well.

I'd probably make a function:

```python
class DockerUrl(NamedTuple):
    host: str
    image: str
    tag: Optional[str] = None


def parse_docker_url(url: str) -> DockerUrl:
    # TODO: implement me!
    ...
```

Reply:
`NamedTuple` class sounds good to me
I just moved and used the `Image` `NamedTuple` I used for testing with a new `_get_image` function that parses a uri as suggested.